### PR TITLE
bpo-36139: release GIL around munmap()

### PR DESCRIPTION
**Pull request by sorcio**
https://bugs.python.org/issue36139

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by benjaminp** [comment]
Should we also be doing this around the `close`?

**Comment by sorcio** [comment]
> Should we also be doing this around the `close`?

It would make sense but I wasn't sure if we should do it separately or in one block, and if it should go inside the `if`s. In my experience `munmap()` is more expensive by far (see issue tracker) so this seemed more simple. We can change it of course.

**Comment by benjaminp** [comment]
I think it makes sense to stick them both in the same block. We're certainly release the GIL when closing files.

**Comment by sorcio** [comment]
@benjaminp I extended the GIL release to include the fd close, both on dealloc and `mmap.close()`. I also included Windows.

File: Modules/mmapmodule.c [hunk 1]
```diff
@@ -180,6 +183,7 @@ mmap_close_method(mmap_object *self, PyObject *unused)
         self->data = NULL;
```
Comment (benjaminp) [inline Modules/mmapmodule.c]: I don't think it's safe to not be holding the GIL here. Someone could be trying to read or write the mmap object while calling `close`.

**Comment by sorcio** [comment]
@benjaminp may I ask for another look? If this is still not ok for `close()` I drop it and only take care of dealloc for now. Or I can use some guidance.

File: Modules/mmapmodule.c [hunk 2]
```diff
@@ -157,28 +159,37 @@ mmap_close_method(mmap_object *self, PyObject *unused)
        again.
        TODO - should we check for errors in the close operations???
     */
-    if (self->data != NULL) {
-        UnmapViewOfFile(self->data);
-        self->data = NULL;
+    HANDLE map_handle = self->map_handle;
+    HANDLE file_handle = self->file_handle;
+    char * data = self->data;
```
Comment (benjaminp) [inline Modules/mmapmodule.c]: `char *data`

File: Modules/mmapmodule.c [hunk 3]
```diff
@@ -157,28 +159,37 @@ mmap_close_method(mmap_object *self, PyObject *unused)
...
+        CloseHandle(file_handle);
     }
+    Py_END_ALLOW_THREADS
 #endif /* MS_WINDOWS */
 
 #ifdef UNIX
-    if (0 <= self->fd)
-        (void) close(self->fd);
+    int fd = self->fd;
+    char * data = self->data;
```
Comment (benjaminp) [inline Modules/mmapmodule.c]: `char *data`

**Review by benjaminp — COMMENTED** [review]
Looks good modulo minor formatting issues.

Please also add a NEWS note.

File: Modules/mmapmodule.c [hunk 4]
```diff
@@ -117,6 +117,7 @@ typedef struct {
 static void
 mmap_object_dealloc(mmap_object *m_obj)
 {
+    Py_BEGIN_ALLOW_THREADS
 #ifdef MS_WINDOWS
     if (m_obj->data != NULL)
         UnmapViewOfFile (m_obj->data);
```
Comment (vstinner) [inline Modules/mmapmodule.c:123]: PyMem_Free(m_obj->tagname) is called below without holding the GIL: that's illegal. Other move the call to free when the GIL is hold again, or use PyMem_RawFree().

"Warning: The GIL must be held when using these functions. "
https://docs.python.org/dev/c-api/memory.html#memory-interface

Note: You can use PYTHONMALLOC=debug or -X dev to reproduce the issue on a Python compiled in release mode.

**Pull request merged**